File: Lab4/fork_and_exec.py
import os
import sys

pid = os.fork()

if pid < 0:
    raise Exception("Fork failed")
elif pid == 0:
    os.execl("/bin/ls", "ls", "-l")
    sys.stderr.write("Exec failed\n")
    sys.exit(1)
else:
    os.waitpid(pid, 0)
    print("Child process completed")

File: Lab4/fork_and_exec.c
#include <stdio.h>
#include <unistd.h>
#include <sys/wait.h>

int main() {
    pid_t pid;

    pid = fork();

    if (pid < 0) {
        perror("Fork failed");
        return 1;
    } else if (pid == 0) {
        // Child process
        char *args[] = {"ls", "-l", "-a", NULL};
        execvp("ls", args);
    } else {
        // Parent process
        wait(NULL);
    }

    return 0;
}